From 4c101d9c70f71358cd61ce39c7d8f2da1bdcfd25 Mon Sep 17 00:00:00 2001
From: Viorel Suman <viorel.suman@nxp.com>
Date: Fri, 9 Feb 2018 11:39:58 +0200
Subject: [PATCH 3354/5242] MLK-17528-3: ASoC: fsl_sai: Set clock rate in
 "set_sysclk" API

commit  4a132f5a1b8dd2da19b3fc86bf58cb13e316778d from
https://source.codeaurora.org/external/imx/linux-imx.git

Set the requested clock rate in "set_sysclk" for specified clock id.

Signed-off-by: Viorel Suman <viorel.suman@nxp.com>
Suggested-by: Shengjiu Wang <shengjiu.wang@nxp.com>
Reviewed-by: Shengjiu Wang <shengjiu.wang@nxp.com>
Reviewed-by: Daniel Baluta <daniel.baluta@nxp.com>
Signed-off-by: Meng Li <Meng.Li@windriver.com>
---
 sound/soc/fsl/fsl_sai.c |   19 +++++++++++++++++++
 1 file changed, 19 insertions(+)

diff --git a/sound/soc/fsl/fsl_sai.c b/sound/soc/fsl/fsl_sai.c
index b4ffa41..52a2ef4 100644
--- a/sound/soc/fsl/fsl_sai.c
+++ b/sound/soc/fsl/fsl_sai.c
@@ -245,6 +245,25 @@ static int fsl_sai_set_dai_sysclk(struct snd_soc_dai *cpu_dai,
 		return 0;
 	}
 
+	if (freq > 0) {
+		if (clk_id < 0 || clk_id >= FSL_SAI_MCLK_MAX) {
+			dev_err(cpu_dai->dev, "Unknown clock id: %d\n", clk_id);
+			return -EINVAL;
+		}
+
+		if (IS_ERR_OR_NULL(sai->mclk_clk[clk_id])) {
+			dev_err(cpu_dai->dev, "Unassigned clock: %d\n", clk_id);
+			return -EINVAL;
+		}
+
+		ret = clk_set_rate(sai->mclk_clk[clk_id], freq);
+		if (ret < 0) {
+			dev_err(cpu_dai->dev, "failed to set clock rate (%u): %d\n",
+					freq, ret);
+			return ret;
+		}
+	}
+
 	ret = fsl_sai_set_dai_sysclk_tr(cpu_dai, clk_id, freq,
 					FSL_FMT_TRANSMITTER);
 	if (ret) {
-- 
1.7.9.5

